### ISSUE TITLE
Analysis of CoAP Implementations for Industrial Internet of Things

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#787]
This could maybe interest some people : 
https://www.researchgate.net/publication/317546381_Analysis_of_CoAP_Implementations_for_Industrial_Internet_of_Things_A_Survey

(tests was made with Cf 1.1.0-SNAPSHOT... :confused: )

**Comment by boaks** [comment]
I stumbled over that some months ago.
Chapter 6, Table 3 focus on RTT.
FMPOV, it is also interessting, how many request a server can process in parallel.
I would guess, that would be more californiums domain.

**Comment by boaks** [comment]
Today was a interesting e-mail from Carsten Bormann in the core - mailing list. 

https://www.zdnet.com/article/the-coap-protocol-is-the-next-big-thing-for-ddos-attacks/

In the middle, under "THE RISE OF COAP", a use statistic is presented.

**Comment by jabberimaher** [comment]
hi, when i launch command mvn package it display:
[INFO] Building mon-appli 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ mon-appli ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/maher/projet1/mon-appli/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ mon-appli ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /home/maher/projet1/mon-appli/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] Source option 5 is no longer supported. Use 6 or later.
[ERROR] Target option 1.5 is no longer supported. Use 1.6 or later.
[INFO] 2 errors 
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.430 s
[INFO] Finished at: 2018-12-22T08:59:08+01:00
[INFO] Final Memory: 9M/37M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project mon-appli: Compilation failure: Compilation failure: 
[ERROR] Source option 5 is no longer supported. Use 6 or later.
[ERROR] Target option 1.5 is no longer supported. Use 1.6 or later.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
how to resolve this problem?

**Comment by boaks** [comment]
I don't know, how you get that aged stuff.
I would not recommend to use the 1.0.x branch for new work. But if, then please update to 1.0.7-SNAPSHOT and it requires Java 1.7.
I would recommend to use the 2.0.x branch for new work.

**Comment by jabberimaher** [comment]
Thanks a lot professor,
i think that the problem is solved.
i modify the pom.xml by adding 
*plugin : 
<configuration>
          <source>1.8</source>
          <target>1.8</target>
        </configuration>
*dependency:
<dependency>
    <groupId>org.eclipse.californium</groupId>
    <artifactId>californium-core</artifactId>
    <version>2.0.0-M12</version>
</dependency>
and i launch mvn package, it dispalys:
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building mon-appli 2.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
Downloading from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/californium-core/2.0.0-M12/californium-core-2.0.0-M12.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/californium-core/2.0.0-M12/californium-core-2.0.0-M12.pom (5.2 kB at 7.3 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/parent/2.0.0-M12/parent-2.0.0-M12.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/parent/2.0.0-M12/parent-2.0.0-M12.pom (16 kB at 17 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/element-connector/2.0.0-M12/element-connector-2.0.0-M12.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/element-connector/2.0.0-M12/element-connector-2.0.0-M12.pom (2.4 kB at 8.5 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/slf4j/slf4j-api/1.7.25/slf4j-api-1.7.25.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/slf4j/slf4j-api/1.7.25/slf4j-api-1.7.25.pom (3.8 kB at 12 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/slf4j/slf4j-parent/1.7.25/slf4j-parent-1.7.25.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/slf4j/slf4j-parent/1.7.25/slf4j-parent-1.7.25.pom (14 kB at 17 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/californium-core/2.0.0-M12/californium-core-2.0.0-M12.jar
Downloading from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/element-connector/2.0.0-M12/element-connector-2.0.0-M12.jar
Downloading from central: https://repo.maven.apache.org/maven2/org/slf4j/slf4j-api/1.7.25/slf4j-api-1.7.25.jar
Downloading from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/element-connector/2.0.0-M12/element-connector-2.0.0-M12-tests.jar
Downloaded from central: https://repo.maven.apache.org/maven2/org/slf4j/slf4j-api/1.7.25/slf4j-api-1.7.25.jar (41 kB at 3.0 kB/s)
Downloaded from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/element-connector/2.0.0-M12/element-connector-2.0.0-M12-tests.jar (81 kB at 4.7 kB/s)
Downloaded from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/element-connector/2.0.0-M12/element-connector-2.0.0-M12.jar (123 kB at 7.0 kB/s)
Downloaded from central: https://repo.maven.apache.org/maven2/org/eclipse/californium/californium-core/2.0.0-M12/californium-core-2.0.0-M12.jar (381 kB at 14 kB/s)
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ mon-appli ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/maher/projet2/mon-appli/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ mon-appli ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ mon-appli ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/maher/projet2/mon-appli/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.0:testCompile (default-testCompile) @ mon-appli ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ mon-appli ---
[INFO] Surefire report directory: /home/maher/projet2/mon-appli/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.exemple.demo.AppTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.017 sec

Results :

Tests run: 1, Failures: 0, Errors: 0, Skipped: 0

[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ mon-appli ---
[INFO] Building jar: /home/maher/projet2/mon-appli/target/mon-appli-2.0-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 33.286 s
[INFO] Finished at: 2018-12-23T08:52:47+01:00
[INFO] Final Memory: 14M/57M
[INFO] ------------------------------------------------------------------------
....
is the procedure followed is just? now i would testing some examples with californium then implementing just the handshake DTLS. how i realize this work and can i implement others authentication algorithms?
i'm sorry Professor to distrub you but may GOD sent you to be the savior of my project, i surf lonely and i need someone guide and advise me.
i am grateful.

**Comment by boaks** [comment]
OK, seems you followed my recommendation to use new stuff, but now your JDK is too new (for M12) :-).

Try to update your POM's  maven-surefire-plugin configuration with
```
<useSystemClassLoader>false</useSystemClassLoader>
```

```
<plugin>
	<artifactId>maven-surefire-plugin</artifactId>
	<version>2.22.1</version>
	<configuration>
		<useSystemClassLoader>false</useSystemClassLoader>
		<systemPropertyVariables>
		...
	</configuration>
</plugin>
```

**Comment by jabberimaher** [comment]
ok, after updating pom.xml by adding:
------------------------------------------------------
<plugin>
<groupId>org.apache.maven.plugins</groupId>
	<artifactId>maven-surefire-plugin</artifactId>
	<version>2.22.1</version>
	<configuration>
		<useSystemClassLoader>false</useSystemClassLoader>
		<systemPropertyVariables></systemPropertyVariables>
		
	</configuration>
</plugin>
--------------------------------------------------------------------------------------------------------
here the result of mvn package:
root@maher-Lenovo-G50-80:/home/maher/projet2/mon-appli# mvn package
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building mon-appli 2.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ mon-appli ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/maher/projet2/mon-appli/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ mon-appli ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ mon-appli ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/maher/projet2/mon-appli/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.0:testCompile (default-testCompile) @ mon-appli ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-surefire-plugin:2.22.1:test (default-test) @ mon-appli ---
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/surefire-junit3/2.22.1/surefire-junit3-2.22.1.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/surefire-junit3/2.22.1/surefire-junit3-2.22.1.pom (2.5 kB at 3.5 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/surefire-providers/2.22.1/surefire-providers-2.22.1.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/surefire-providers/2.22.1/surefire-providers-2.22.1.pom (2.5 kB at 17 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/surefire-junit3/2.22.1/surefire-junit3-2.22.1.jar
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/surefire-junit3/2.22.1/surefire-junit3-2.22.1.jar (27 kB at 121 kB/s)
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.exemple.demo.AppTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 s - in org.exemple.demo.AppTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ mon-appli ---
[INFO] Building jar: /home/maher/projet2/mon-appli/target/mon-appli-2.0-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.599 s
[INFO] Finished at: 2018-12-23T17:55:32+01:00
[INFO] Final Memory: 14M/57M
[INFO] ------------------------------------------------------------------------
root@maher-Lenovo-G50-80:/home/maher/projet2/mon-appli#

**Comment by jabberimaher** [comment]
<plugin>
<groupId>org.apache.maven.plugins</groupId>
	<artifactId>maven-surefire-plugin</artifactId>
	<version>2.22.1</version>
	<configuration>
		<useSystemClassLoader>false</useSystemClassLoader>
		<systemPropertyVariables></systemPropertyVariables>
		
	</configuration>
</plugin>

**Comment by boaks** [comment]
If it builds, you may use `demo-apps/sc-dtls-example-server` as server and `demo-apps/sc-dtls-example-client` as client. To do so, open a terminal in the californium/demo-apps/run folder and execute `java -jar sc-dtls-example-server-2.0.0-SNAPSHOT.jar` and afterwards from an other terminal in that folder `java -jar sc-dtls-example-client-2.0.0-SNAPSHOT.jar`.
The client should the print

> Create 1 DTLS example clients, expect to send 100 messages overall.
> 1 DTLS example clients started.
> 1 DTLS example clients finished.
> 100 messages received, 100 expected
> 100 messages in 338 ms
> 295 messages per s

Or something similar.

> how i realize this work and can i implement others authentication algorithms?

This depends of the scope of your "authentication algorithms" and your work.
 
If your intention is to "authenticate" the other peer, californium offers ciphersuites with PSK (PreSharedKey), Raw Public Key, x509v3 Certificate based ciphersuites.

TLS_PSK_WITH_AES_128_CBC_SHA256
TLS_ECDHE_PSK_WITH_AES_128_CBC_SHA256
TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
TLS_PSK_WITH_AES_128_CCM_8
TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8

So, if you want to compare them, you may use demo-apps/cf-secure project.
Also open two terminals in californium/demo-apps/run and start the server with

> java -jar cf-secure-2.0.0-SNAPSHOT.jar

and the client with

> java -cp cf-secure-2.0.0-SNAPSHOT.jar org.eclipse.californium.examples.SecureClient PSK

At the end of the console output, there should be something as:
 
```
==[ CoAP Response ]============================================
MID    : 27727
Token  : 746DFCEF9FACAC46
Type   : ACK
Status : 2.05
Options: {"Content-Format":"text/plain"}
RTT    : 129 ms
Payload: 14 Bytes
---------------------------------------------------------------
hello security
===============================================================
```

You may replace PSK with RPK or X590.

If you want to compare with other ciphersuites, e.g. some with MD5, I would recommend to use some TLS (openssl) libraries, which offers that. To extend californium with old MD5 is not planned at all and will be extremely hard to do.

I' still not understand, why you want to use java to measure the energy consumption.

**Comment by boaks** [comment]
I feel, that there potentially misunderstandings:

If you just want to compare MD5 with SHA, you don't need TLS/DTLS, just use these basic crypto-functions on there own. That would also be true, for other basic crypto-functions e.g. AES compared with ChaCha or 3DES.

If you want to compare DTLS/TLS ciphersuites (a set of basic crypto-functions to secure data within communication), I would go for openssl or gnuTLS, they offer a broad collection of such suites. To implement somethings as a "new cipher suite" is always hard, in californium, but also in openssl. So, I'm not sure, what the scope of your analysis should be.

**Comment by jabberimaher** [comment]
Thanks Professor for your interest
as part of the preparation of a first paper in my thesis project, it is asked to compare between 4 authentication algorithms that are already created by other researchers (dtls handshake, Message Digest Based Authentication (MDA), MDLA, .. .), the comparison will touch the metrics of performaces (energy, time of execution, throughput, ...).
this comparison helps us to fix what is the most reliable algorithm for sensor nodes in a wsn.
So I'm not going to create a new algorithm, it might be in an advanced internship in my thesis project.
then the work is described as follows:
-Search and locate algorithms that can be used in a sensor-cloud environment.
(already realized: DTLS, MDA, MDLA).
-implement DTLS handshake with a suitable simulator (contiki, californium, ...)
-inspire from the implementation of DTLS handshake an implementation of the other algorithms already chosen)
for each implemented algorithm use the tools of the simulator used to derive measurement results of performance metrics (energy, execution time, throughput, etc.)
is this work feasible with californium.
best regards

**Comment by jabberimaher** [comment]
this is a part of MDA
_**Algorithm 1 Registration: mobile_with_cloud**_
Require:
isAlive (mobile, cloud)
hasNetworkAccess (mobile)
Ensure:
Role_Mobile
const userID
var password
uid = hash(userID)
pwd = hash(password)
cloud ← send(uidpwd)
Tk = uid ⊕ pwd
Role_Cloud
recv(uidpwd)
Tk = uid ⊕ pwd
EXP = Registration expiry period for the client
#CF ← pointer{store(uid, pwd, Tk)}
MDuser = hash(usage policy, user access level, user
certificate)
MDcloud = hash(user add policy, cloud resource
restriction, cloud certificate)
MD = hash(MDcloudMDuser)
Temp = MDuserMDcloud#CFPkpub_cloud
msz = encrypt(Temp, T)
mobile ← send(msz)
**_Algorithm 2 Mobile_to_cloud_auth_req_**
Require:
isAlive (mobile, cloud)
hasNetworkAccess (mobile)
isRegistered (mobile, cloud)
uid = hash(userID)
pwd = hash(password)
cloud ← send(uidpwd)
Tk = uid ⊕ pwd
Ensure:
MD = hash(MDcloudMDuser)
Auth_Keyi ← Tk PRNG SI
Enc1 = encrypt(MD,Auth_Keyi)
Temp = Enc1SI
Enc2 = encrypt(Temp, Tk)
cloud ← send(Enc2#CF)

**Comment by jabberimaher** [comment]
i still wait for your answer Professor
best regards

**Comment by boaks** [comment]
It's still Christmas and I'm not in the office :-).

**Comment by boaks** [comment]
So, are you now able to build californium 2.0.0-M12 or SNAPSHOT?
If yes, are you able to execute the demo examples as I recommended above?
If so, may be you can give other a feedback, what you did to make it working. 
I would really prefer, if that build issues are focused on your other issue #833.
So please, add you solution there and then close that issue.

**Comment by boaks** [comment]
About the possibilities to sue Californium for MDA:

You may just use the possibility to transfer "binary payload (byte arrays)" in requests and responses, but there will not be to much benefit for your analyses doing so with CoAP compared to plain UDP. It's up to you. But trying to implement MDA "somehow" with Californium, is not really related with this issue. I recommended you to read the referred paper to get some inspiration. So, if you want to reach other californium user with interesst in implementing MDA on top of californium, please open a new issue for that topic,

**Comment by boaks** [comment]
If you put the "build issue stuff" in #833 and create a new issue for the MDA, I would like to delete our comments on this issue, because they are not really related here. WOuld this be OK for you?

**Comment by boaks** [comment]
Please move the "build" discussion to your issue #833, as already recommended.

I will not longer answer build question in this "Analysis of CoAP Implementations for Industrial Internet of Things" issue, these issues are not related!

**Comment by boaks** [comment]
@sbernard31 

I moved the links to a wiki page.
[Links to CoAP   DTLS 1.2 research information](https://github.com/eclipse/californium/wiki/Links-to-CoAP---DTLS-1.2-research-information)

If you OK with that, please close this issue.

**Comment by sbernard31** [comment]
:+1: 
If you add more resources to this page, please let me know.

**Comment by boaks** [comment]
I added the RFC and some lucky links.
